### PR TITLE
fix(babel): keep function names to improve stack traces in tests

### LINKED ISSUE #188: babel-plugin-minify-dead-code-elimination removes function names
- `kcd-scripts` version: 7.5.3
- `node` version: 12.20.0
- `npm` (or `yarn`) version: 6.14.8

Relevant code or config

```javascript
export function createFunc1() {
  function func() {
  }
  return func
}

export function createFunc2() {
  function func() {
  }
  const name = func.name // not returned, just used
  return func
}
```

What you did:

Run `kcd-scripts build`

What happened:

The output is

```javascript
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createFunc1 = createFunc1;
exports.createFunc2 = createFunc2;

function createFunc1() {
  return function () {};
}

function createFunc2() {
  function func() {}

  func.name; // not returned, just used

  return func;
}
```

Reproduction repository:

[Babel REPL](https://babeljs.io/en/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=usage&spec=false&loose=false&code_lz=KYDwDg9gTgLgBAMwK4DsDGMCWEVzVYAQxmADFU0BGACgEo4BvAKDkQqxzfTsYF8mWcAjCRRcydAG4m_UJFhcM2XPiIly6AEw9mrCUs76d_QWhwBneCkIBbYHAC8igHTW7kuAHpPcFBHjCoijAACYANHAAVkiWcDGhAqyBYorSvEA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ces2015%2Creact%2Ctypescript&prettier=true&targets=&version=7.12.12&externalPlugins=babel-plugin-minify-dead-code-elimination%400.5.1)

Problem description:

The name of the function returned in `createFunc1` gets removed thanks to `babel-plugin-minify-dead-code-elimination`.  This is causing me problems in a node package (specifically a testing library) where the name gets removed and consequently does not appear in stack traces which hurts debugability when things go wrong.

I can understand the desire to remove this when building for browsers (reduced bundle size), but I don't personally see many advantages to this for a node/test package.

Suggested solution:

There is an [option to keep function names in the plugin](https://github.com/babel/minify/tree/master/packages/babel-plugin-minify-dead-code-elimination#options), so we could set that the be true and keep the rest of the functionality.  We could use the [`isTest` flag](https://github.com/kentcdodds/kcd-scripts/blob/2e7e31267d1fab9dcb4d9a0d3a589cb8b3726472/src/config/babelrc.js#L14) to conditionally set the option for tests only.

Similarly, we could use the `isTest` flag to remove the plugin all together for tests.

Either solution would suffice for my issue.  I'm also happy to submit the PR for this change if you agree with it and can advise me which approach is your preference.

### PR DESCRIPTION
**Pull request by mpeyper**
**What**:

Preserve function names when eliminating dead code in tests to enable nicer stack traces when errors occur.

**Why**:

Fixes #188

**How**:

I used the `isTest` flag to enable [the `keepFnName` option of `babel-plugin-minify-dead-code-elimination`](https://github.com/babel/minify/tree/master/packages/babel-plugin-minify-dead-code-elimination#options).

**Checklist**:

- [ ] Documentation N/A
- [ ] Tests N/A
- [x] Ready to be merged

### CODE REVIEW

**Comment by mpeyper** [comment]
I'm having some second thoughts on this. This will only help the case where a project is using `kcd-scripts` and running their own tests to see the function names of their own code.  For my case, I'm writing a library to be used in tests, so if I build my code with `kcd-scripts` to publish to `npm` the function names will still be removed.

Digging through the history a bit, it looks like this plugin came about [as part of an UMD feature](https://github.com/kentcdodds/kcd-scripts/commit/49e757e7409788bc8b2db173cd624187dac03327), which sort of makes sense to me as I'd expect the consumer of the package to deal with the bundling, minifying, treeshaking, etc. when using it from NPM.  I'm wondering if should only be added to the config if:

1. `isUMD` is `true`
2. and/or `BUILD_MINIFY` environment variable is set to `true`

I've also discovered that this issue is killing the display names of React HOCs:

```jsx
const withValue = (value) => (Component) => {
  const WithValue = (props) => (
    <Component {...props} value={value} />
  )
  return WithValue
}
```

becomes:

```jsx
const withValue = value => Component => {
  return props => /*#__PURE__*/React.createElement(Component, (0, _extends2.default)({}, props, {
    value: value
  }));
};
```

(although it could be argued that HOCs should be setting their display name explicitly).

**Comment by kentcdodds** [comment]
Yeah, let's only enable the minification plugins when `BUILD_MINIFY` is true 👍

File: src/config/babelrc.js [hunk 1]
```diff
@@ -81,7 +81,7 @@ module.exports = () => ({
       ? require.resolve('babel-plugin-transform-inline-environment-variables')
       : null,
     [require.resolve('@babel/plugin-proposal-class-properties'), {loose: true}],
-    require.resolve('babel-plugin-minify-dead-code-elimination'),
+    ['babel-plugin-minify-dead-code-elimination', {keepFnName: isTest}],
```
Comment (kentcdodds) [inline src/config/babelrc.js:84]: Let's add:

```js
const isMinify = parseEnv('BUILD_MINIFY', false)
```

And then change this to:

```suggestion
    isMinify ? require.resolve('babel-plugin-minify-dead-code-elimination') : null,
```

Comment (mpeyper) [reply src/config/babelrc.js]: Done.  Also, I didn't even notice I have removed the `require.resolve` in my first commit 😅 

P.S. I would have taken your suggestion if it were possible to add the `parseEnv` at the same time.

**Review by kentcdodds — APPROVED** [review]
Awesome, thanks 👍

**Comment by kentcdodds** [comment]
@All-contributors please add @mpeyper for code

**Pull request merged**